java/ql/src/Security/CWE/CWE-297/UnsafeHostnameVerification.qhelp

```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE qhelp PUBLIC
+  "-//Semmle//qhelp//EN"
+  "qhelp.dtd">
+<qhelp>
+<overview>
+<p>
+If a <code>HostnameVerifier</code> always returns <code>true</code> it will not verify the hostname at all.
+This stops Transport Layer Security (TLS) providing any security and allows an attacker to perform a Man-in-the-middle attack against the application.
+</p>
+
+<p>
+An attack might look like this:
+</p>
+
+<ol>
+  <li>The program connects to <code>https://example.com</code>.</li>
+  <li>The attacker intercepts this connection and presents one of their valid certificates they control, for example one from Let's Encrypt.</li>
+  <li>Java verifies that the certificate has been issued by a trusted certificate authority.</li>
+  <li>Java verifies that the certificate has been issued for the host <code>example.com</code>, which will fail because the certificate has been issued for <code>malicious.domain</code>.</li>
+  <li>Java wants to reject the certificate because the hostname does not match. Before doing this it checks whether there exists a <code>HostnameVerifier</code>.</li>
+  <li>Your <code>HostnameVerifier</code> is called which returns <code>true</code> for any certificate so also for this one.</li>
```

Comment:
We normally talk about code elements as "The <code>example</code> class"  (or method, interface, function...). Forgive my ignorance, but I'm not sure what we should use here.

Reply:
I intended this to represent how the attack would work in practice.
That's why it was numbered.

What do you think of 
```suggestion
  <li>The insecure <code>HostnameVerifier</code> is called which returns <code>true</code> for any certificate so also for this one.</li>
```
?

Reply:
I asked about this and it sounds as if I was thinking about `HostnameVerifier` the wrong way - my apologies. I'm not a writer, not a developer, so the informal "rules" that I've built up don't always hold. 😞 

I'll resolve this conversation.

Reply:
I'm always happy when you or another writer checks my writing :)